### PR TITLE
plugincontainer: Support mlock

### PR DESCRIPTION
**Pull request by tomhjp**
When Vault's [disable_mlock](https://developer.hashicorp.com/vault/docs/configuration#disable_mlock) config setting is not set, it also tells plugins to call `unix.Mlockall` to ensure they don't swap any memory (with potentially sensitive data) out to disk. Without the IPC_LOCK capability, this is not possible inside a container, and previously meant containerized plugins would fail to start. https://github.com/hashicorp/vault/pull/23215 is a WIP PR with the consumption of the new `Config.CapIPCLock` option.

To ensure better test coverage, I formalised the options we are interested in into a TestCompatibilityMatrix test. Some tests are skipped as they are known not to work, and some are skipped as TODO items, waiting to set up CI so they can pass. I'll be working to remove most/all of these skips over the coming weeks.

Also fixes a couple of small bugs and tidy-ups:
* Config's Tag field was not correctly set, found in one of the new tests
* The error message for a plugin that needs updating reported one of the wrong directories
* Refactors most of the tests into an external `plugincontainer_test` package so they are forced to use the library like an external consumer (Vault) would.
* A couple of errors are made more descriptive.

### CODE REVIEW

File: plugincontainer/compatibility_matrix_test.go [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+		s += ":" + "nonroot"
+	}
+	if m.mlock {
+		s += ":" + "mlock"
+	}
+	return s
+}
+
+func TestCompatibilityMatrix(t *testing.T) {
+	runCmd(t, "go", "build", "-o=examples/container/go-plugin-counter", "./examples/container/plugin-counter")
```
Comment (swenson) [inline plugincontainer/compatibility_matrix_test.go:49]: Should we skip the tests on macOS? Currently I get a failure for these with something like `Did not find expected socket file: stat /run/user/1000/docker.sock: no such file or directory`

**Review by swenson — COMMENTED** [review]
Looks mostly good. One question in the `main()`.

I appreciate all of the automated tests, as it is a pain to get all of the combinations working on my own Linux desktop :)

File: plugincontainer/compatibility_matrix_test.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+	for _, engine := range []string{engineDocker, enginePodman} {
```
[the remark this answers is not in the document]
Comment (tomhjp) [reply plugincontainer/compatibility_matrix_test.go:51]: Haha yeah it also doesn't feel very sustainable if I need to add more combinations in future. Will have a think about whether there are ways to do this more elegantly.

**Review by swenson — APPROVED** [review]
LGTM

**Comment by tomhjp** [comment]
Thanks!

**Pull request merged**